enhancements/network/user-space-pod-iface-and-api-library.md

```diff
@@ -0,0 +1,382 @@
+application.
+
+- Ease the consumption of the data that was passed into a user space networking
+application by providing a library that both provides an example of how the
+data can be collected and if desired, integrated directly into the application. 
+
+- Ensure customers using SR-IOV can run user space networking applications in
+their deployment.
+
+### Non-Goals
+
+The following items are out of scope for this proposal:
+
+- This could, secondarily, also be used to support user space data plane
+technologies like ovs-dpdk, but its not the same, and its not our primary goal.
+
+- Non-SR-IOV devices (vDPA and vhost-user interfaces) require additional
+information. However, these devices are not required for this enhancement.
+That being said, some of the upstream work this enhancement is pulling from
+does include those interfaces. So the API Library may include some of the
+additional work for these interfaces.
+
+## Proposal
+
+As described above, there are two parts to this enhancement. Work is need to
+pass additional data to the pod to enable user space networking applications to
+run properly. Work is also needed in the pod to ease the consumption of the
+data. Each item will be described below.
+
+### Passing Data to the Pod
+
+Currently, data is passed to the pod in several forms:
+
+- **PCI Addresses**: SR-IOV Device Plugin passes the list of PCI Addresses in
+an environment variable to the pod.
+
+- **Interface Data**: Multus populates the pod annotation with a
+`k8s.v1.cni.cncf.io/network-status` field. This annotation contains the list
+of interfaces, and for each interface, the MAC address, list of IP addresses
+(not CIDR), interface (net1, net2, etc), name, etc. The pod can use the
+Kubernetes Downward API to retrieve this information.
+
+- **CPU List**: The dedicated CPU list can be retrieved by reading
+`/sys/fs/cgroup/cpuset/cpuset.cpus` in the pod.
+
+For user space networking applications running in a pod, the following
+information is still needed:
+
+- **Hugepage Information**: The type and number of hugepages allocated to the
+pod.
+
+- **Additional Interface Data**: The `k8s.v1.cni.cncf.io/network-status`
+contains a `name` field, which is the network in the
+[Network-Attachment-Definition](#networkattachmentdefinition). The `name`
+does indicate how the interface is intended to be used, however, the pod doesn't
+have enough information to determine which PCI Address is associated with which
+interface. So additional information is needed to make that association within
+the pod.
+
+#### Hugepage Information
+
+There was an upstream Kubernetes PR,
+[86102](https://github.com/kubernetes/kubernetes/pull/86102),
+to enhance the Kubernetes Downward API to add hugepage information to the set
+of supported fields. This PR was closed due to inactivity. The plan is to
+resubmit the PR and try to push the functionality into Kubernetes. It doesn't
+appear that there were any major objections to the PR, just lack of activity.
```

Comment:
This paragraph is already out of date, as it looks like @derekwaynecarr reopened the PR last week

Reply:
@danwinship Thanks for looking again. I'll address your comments shortly.

I have been attending sig-node meetings to determine how to address the hugepage issue which prompt Derek reopen. I'll cleanup and make the text more generic.

With regards to your other comment about using the name to determine how an interface is used, I agree it is not great and this is the only reason I have not pushed to get the document merged. I talked with Zenghui some on this and we haven't come up with a better solution. If you have any recommendations let us know.